cli/tests/watchdog.bats

```diff
@@ -87,49 +87,70 @@ process_compose_pids_called_with_arg() {
   echo "$pids"
 }
 
+# Writes the PIDs of all watchdogs for the current project to `filename`.
+write_watchdog_pids_to_file() {
+  filename="$1"
+  watchdog_pids_called_with_arg "$PROJECT_DIR" > "$filename"
+}
+
+# Verifies that `filename` contains a single PID. Does not verify that the PID
+# is still running or was of a watchdog, because it will have exited.
+#
+# The use of `wait_for_watchdogs` in `teardown` verifies that a watchdog is no
+# longer running when the test ends.
+verify_watchdog_pid_file() {
+  filename="$1"
+  pid="$(cat "$filename")"
+  if ! [[ "$pid" =~ ^[0-9]+$ ]]; then
+    echo "Expected a single PID, got '$pid'" >&2
+    exit 1
+  fi
+}
+
 # ---------------------------------------------------------------------------- #
 
 @test "watchdog: can run flox-watchdog" {
   run "$WATCHDOG_BIN" --help
   assert_success
 }
 
-@test "watchdog: lives as long as the activation" {
+@test "watchdog: lives as long as the activation (command)" {
   setup_sleeping_services
-  export -f watchdog_pids_called_with_arg
-  SHELL="bash" run --separate-stderr "$FLOX_BIN" activate -- bash <(cat <<'EOF'
+  export -f write_watchdog_pids_to_file watchdog_pids_called_with_arg
+  SHELL="bash" run "$FLOX_BIN" activate -- bash -c '
+    write_watchdog_pids_to_file "watchdog_pids"
+  '
+  assert_success
+  verify_watchdog_pid_file "watchdog_pids"
+}
 
-    # Ensure that the watchdog is still running
-    times=0
-    while true; do
-      if [ "$times" -gt 100 ]; then
-        exit 1
-      fi
-      pid="$(watchdog_pids_called_with_arg "$_FLOX_SERVICES_SOCKET")"
-      if [ -n "${pid?}" ]; then
-        echo "$pid"
-        break
-      fi
-      times=$((times + 1))
-      sleep 0.01
-    done
-EOF
-)
-  pid="$output"
+@test "watchdog: lives as long as the activation (in-place)" {
+  setup_sleeping_services
+  export -f write_watchdog_pids_to_file watchdog_pids_called_with_arg
+  run bash -c '
+    set -euo pipefail
+    eval "$("$FLOX_BIN" activate)"
+    write_watchdog_pids_to_file "watchdog_pids"
+  '
   assert_success
+  verify_watchdog_pid_file "watchdog_pids"
+}
 
-  # Ensure that the watchdog has exited now
-  times=0
-  while true; do
-    if [ "$times" -gt 100 ]; then
-      exit 1
-    fi
-    if ! kill -0 "$pid"; then
-      break
-    fi
-    times=$((times + 1))
-    sleep 0.01
-  done
+# evaling the activation can be deferred up to the 5s timeout in `flox-activations attach`.
+@test "watchdog: lives as long as the activation (in-place deferred)" {
+  setup_sleeping_services
+  export -f write_watchdog_pids_to_file watchdog_pids_called_with_arg
+  run bash -c '
+    set -euo pipefail
+    "$FLOX_BIN" activate > to_eval
+    write_watchdog_pids_to_file "watchdog_pids1"
+    eval "$(cat to_eval)"
+    write_watchdog_pids_to_file "watchdog_pids2"
+  '
+  assert_success
+  verify_watchdog_pid_file "watchdog_pids1"
+  verify_watchdog_pid_file "watchdog_pids2"
+  [ "$(cat watchdog_pids1)" = "$(cat watchdog_pids2)" ]
```

Comment:
nit: I agree the value of this test is a bit questionable and wonder if it might just be better to drop. Even if we did regress, I wouldn't count on this test to catch it because there would probably be a race between the watchdog and the `eval "$(cat to_eval)"`

Reply:
Yeah, I'll drop it. Any preference about keeping "watchdog: lives as long as the activation (in-place)"?

Reply:
I lean towards dropping. Same for `watchdog: lives as long as the activation (command)`. The services tests already cover this (e.g. status: `lists the statuses for services`). And this is a racey way to assert that something isn't happening. If we need more coverage we could add a unit test similar to `pid_not_pruned_before_expiration`. But I think the coverage in the services tests is probably sufficient

Reply:
Omitted the new tests and removed the existing test with the description:

> This test only guarantees that the watchdog has been started. Not that
> it lives precisely as long as the activation, which is very hard to do.
> 
> We have better coverage in:
> 
> - services tests; which interact with services from within an activation
> - wait_for_watchdogs; which asserts that we don't leave a watchdog behind
> - unit and integration tests; which assert that the watchdog exits when
>   PIDs are no longer running